File: sfml-quads/Ray.cpp
#include "commons.hpp"
#include "Ray.hpp"
#include "Player.hpp"

#include "SelbaWard/Line.hpp"

#include <iostream>
#include <cmath>
#include <vector>


void Ray::cast(float f, Player &player, std::shared_ptr<Wall> map, std::shared_ptr<sf::RenderWindow> window) {
    static constexpr int texWidth = 64;
    static constexpr int texHeight = 64;
   
    int buffer[WIDTH][HEIGHT]; // y-coordinate first because it works per scanline
    std::vector<int> texture[8];
    for(int i = 0; i < 8; i++) texture[i].resize(texWidth * texHeight);

    for(int x = 0; x < texWidth; x++)
    for(int y = 0; y < texHeight; y++) {
        int xorcolor = (x * 256 / texWidth) ^ (y * 256 / texHeight);
        //int xcolor = x * 256 / texWidth;
        int ycolor = y * 256 / texHeight;
        int xycolor = y * 128 / texHeight + x * 128 / texWidth;
        texture[0][texWidth * y + x] = 65536 * 254 * (x != y && x != texWidth - y); //flat red texture with black cross
        texture[1][texWidth * y + x] = xycolor + 256 * xycolor + 65536 * xycolor; //sloped greyscale
        texture[2][texWidth * y + x] = 256 * xycolor + 65536 * xycolor; //sloped yellow gradient
        texture[3][texWidth * y + x] = xorcolor + 256 * xorcolor + 65536 * xorcolor; //xor greyscale
        texture[4][texWidth * y + x] = 256 * xorcolor; //xor green
        texture[5][texWidth * y + x] = 65536 * 192 * (x % 16 && y % 16); //red bricks
        texture[6][texWidth * y + x] = 65536 * ycolor; //red gradient
        texture[7][texWidth * y + x] = 128 + 256 * 128 + 65536 * 128; //flat grey texture
    }







    float dist = 0.0f;

    int dof = 0, fov = f;
    float rayX = 0.0f, rayY = 0.0f, rayAngle = 0.0f;
    float rayStep = 1.0f;

    int cell = map->cell(), mapW = map->width(), mapH = map->height(), mapTV = 0, mapTH = 0;
    int mapX = 0.0f, mapY = 0.0f, mapPos = 0.0f;
    float offsetX = 0.0f, offsetY = 0.0f;
    std::vector<int> mapGrid = map->grid();

    float playerX = player.position().x;
    float playerY = player.position().y;
    float playerAngle = hkk::radians(player.rotation());

    // Ray angle in radians
    rayAngle = playerAngle - hkk::radians(fov/2);
    if(rayAngle < 0)      rayAngle += 2*M_PI;
    if(rayAngle > 2*M_PI) rayAngle -= 2*M_PI;

    for(float r = 0; r < fov; r += rayStep) {
        // Horizontal line
        float distH = INFINITY;
        float horX = playerX, horY = playerY;

        dof = 0;
        float aTan = -1/tan(rayAngle);


        if(rayAngle > M_PI) {
            rayY = (((int)playerY/cell)*cell) - 0.0001;     // Rounding to closest cell
            rayX = (playerY-rayY)*aTan+playerX;
        
            offsetY = -cell;
            offsetX = -offsetY*aTan;
        }
        if(rayAngle < M_PI) {
            rayY = (((int)playerY/cell)*cell) + cell;       // Rounding to closest cell
            rayX = (playerY - rayY)*aTan+playerX;

            offsetY =  cell;
            offsetX = -offsetY*aTan;
        }
        if(rayAngle == 0 || rayAngle == M_PI) {
            rayX = playerX;
            rayY = playerY;
            dof = 8;
        }


        while(dof < 8) {
            mapX = (int)(rayX)/cell;
            mapY = (int)(rayY)/cell;
            mapPos = mapY * mapW + mapX;

            if((mapPos > 0) && (mapPos < mapW * mapH) && (mapGrid[mapPos] > 0)) {
                mapTH = mapGrid[mapPos];
                horX = rayX;
                horY = rayY;
                distH = hkk::dist(playerX, playerY, horX, horY);

                dof = 8;
            } else {
                rayX += offsetX;
                rayY += offsetY;
                dof += 1;
            }
        }


        // Vertical line
        float distV = INFINITY;
        float vertX = playerX, vertY = playerY;

        dof = 0;
        float nTan = -tan(rayAngle);


        if(rayAngle > M_PI_2 && rayAngle < 3*M_PI_2) {
            rayX = (((int)playerX/cell)*cell) - 0.0001;
            rayY = (playerX-rayX)*nTan+playerY;

            offsetX = -cell;
            offsetY = -offsetX*nTan;
        }
        if(rayAngle < M_PI_2 || rayAngle > 3*M_PI_2) {
            rayX = (((int)playerX/cell)*cell) + cell;
            rayY = (playerX-rayX)*nTan+playerY;

            offsetX =  cell;
            offsetY = -offsetX*nTan;
        }
        if(rayAngle == 0 || rayAngle == M_PI) {
            rayX = playerX;
            rayY = playerY;
            dof = 8;
        }


        while(dof < 8) {
            mapX = (int)(rayX)/cell;
            mapY = (int)(rayY)/cell;
            mapPos = mapY * mapW + mapX;

            if((mapPos > 0) && (mapPos < mapW * mapH) && (mapGrid[mapPos] > 0)) {
                mapTV = mapGrid[mapPos];
                vertX = rayX;
                vertY = rayY;
                distV = hkk::dist(playerX, playerY, vertX, vertY);

                dof = 8;
            } else {
                rayX += offsetX;
                rayY += offsetY;
                dof += 1;
            }
        }

        int shade = 0;
        sf::Color wallpaint;
        
        float distSqr = 0.0f;
        float brightness = 0.0f;
        float widthSqr = (WIDTH*WIDTH);

        if(distV < distH) {
            rayX = vertX;
            rayY = vertY;
            dist = distV;


            distSqr = (dist*dist);
            brightness = hkk::map(distSqr, 0.0f, widthSqr, 255.0f, 0.0f);     // Brightness based wall shading
        
            switch(mapTV) {
                case 1: wallpaint.r = brightness; break;
                case 2: wallpaint.g = brightness; break;
                case 3: wallpaint.b = brightness; break;
                case 4: {wallpaint.r = brightness; wallpaint.g = brightness;} break;
                default: break;
            }
        }
        if(distH < distV) {
            rayX = horX;
            rayY = horY;
            dist = distH;


            distSqr = (dist*dist);
            brightness = hkk::map(distSqr, 0.0f, widthSqr, 230.0f, 0.0f);     // Brightness based wall shading
        
            switch(mapTH) {
                case 1: wallpaint.r = brightness; break;
                case 2: wallpaint.g = brightness; break;
                case 3: wallpaint.b = brightness; break;
                case 4: {wallpaint.r = brightness; wallpaint.g = brightness;} break;
                default: break;
            }
        }

        window->draw(sw::Line(sf::Vector2f(playerX, playerY), sf::Vector2f(rayX, rayY), 0.0f, wallpaint));

        rayAngle += hkk::radians(rayStep);  // Move one ray radian(rayStep) from another

        if(rayAngle < 0)      rayAngle += 2*M_PI;
        if(rayAngle > 2*M_PI) rayAngle -= 2*M_PI;


        // 3D walls
        // Removes fish eye effect
        float cellAngle = playerAngle - rayAngle;
        if(cellAngle < 0)       cellAngle += 2*M_PI;
        if(cellAngle > 2*M_PI)  cellAngle -= 2*M_PI;
        dist *= cos(cellAngle);


        float lineH = (cell*400)/dist;
        // if(lineH > 320) lineH = 320;

        // Set brightness between bounds
        if(brightness < 0)   brightness = 0;
        if(brightness > 255) brightness = 255;

        float wallWidth = WIDTH/cell;
        float lineOffset = 240 - lineH/2;

        // Draw walls with thick lines
        sf::Texture wallTex;
            wallTex.loadFromFile("test_pattern1.bmp");

        sw::Line wall;
            wall.setThickness(wallWidth);
            wall.setPoint(wall.getStartIndex(), {r*wallWidth+530, lineOffset});
            wall.setPoint(wall.getEndIndex()  , {r*wallWidth+530, lineH+lineOffset});
            // wall.setColor(sf::Color(brightness, brightness, brightness, 255));
            wall.setColor(wallpaint);
            wall.setTexture(wallTex);
        window->draw(wall);
    }
}









    // const int mapWidth = 24;
    // const int mapHeight = 24;

    // int worldMap[mapWidth][mapHeight]= {
    //     {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,2,2,2,2,2,0,0,0,0,3,0,3,0,3,0,0,0,1},
    //     {1,0,0,0,0,0,2,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,2,0,0,0,2,0,0,0,0,3,0,0,0,3,0,0,0,1},
    //     {1,0,0,0,0,0,2,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,2,2,0,2,2,0,0,0,0,3,0,3,0,3,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,4,4,4,4,4,4,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,0,4,0,0,0,0,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,0,0,0,0,5,0,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,0,4,0,0,0,0,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,0,4,4,4,4,4,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,4,4,4,4,4,4,4,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
    //     {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1},
    // };

    // sf::Vector2f playerPos = player._player.getPosition();
    // float posX = playerPos.x;
    // float posY = playerPos.y;

    // float dirX = hkk::fromAngle(hkk::radians(player.rotation())).x;
    // float dirY = hkk::fromAngle(hkk::radians(player.rotation())).y;

    // float planeX = hkk::perpendicular(playerPos).x;
    // float planeY = hkk::perpendicular(playerPos).y - 0.34;

    // for(int x = 0; x < WIDTH; x++) {
    //     float cameraX = 2*x / float(WIDTH) - 1;
    //     float rayDirX = dirX + planeX * cameraX;
    //     float rayDirY = dirY + planeY * cameraX;

    //     // int mapX = int(posX)/30;
    //     // int mapY = int(posY)/30;
    //     // int mapPos = mapY * mapWidth + mapX;

    //     int mapX = int(posX/30);
    //     int mapY = int(posY/30);

    //     float sideDistX;
    //     float sideDistY;

    //     float deltaDistX = (rayDirX == 0) ? INFINITY : std::abs(1/rayDirX);
    //     float deltaDistY = (rayDirY == 0) ? INFINITY : std::abs(1/rayDirY);
    //     float perpWallDist;

    //     int stepX;
    //     int stepY;

    //     int hit = 0;
    //     int side;

    //     if(rayDirX < 0) {stepX = -1; sideDistX = (posX - mapX) * deltaDistX;}
    //     else {stepX = 1; sideDistX = (mapX +1.0 - posX) * deltaDistX;}

    //     if(rayDirY < 0) {stepY = -1; sideDistY = (posY - mapY) * deltaDistY;}
    //     else {stepY = 1; sideDistY = (mapY + 1.0 - posY) * deltaDistY;}
        
    //     while(hit == 0) {
    //         if(sideDistX < sideDistY) {
    //             sideDistX += deltaDistX;
    //             mapX += stepX;
    //             side = 0;
    //         } else {
    //             sideDistY += deltaDistY;
    //             mapY += stepY;
    //             side = 1;
    //         }

    //         if(worldMap[mapX][mapY] > 0) hit = 1;
    //     }

    //     // Calculating distance from camera plane - no need to "fisheye" correction
    //     if(side == 0) perpWallDist = (sideDistX - deltaDistX);
    //     else          perpWallDist = (sideDistY - deltaDistY);

    //     int lineHeight = (int)(HEIGHT / perpWallDist);
        
    //     int drawStart = -lineHeight / 2+HEIGHT / 2;
    //     if(drawStart < 0) drawStart = 0;

    //     int drawEnd = lineHeight / 2+HEIGHT / 2;
    //     if(drawEnd >= HEIGHT) drawEnd = HEIGHT-1;

    //     sf::Color wallpaint;
    //     switch(worldMap[mapX][mapY]) {
    //         case 1:  wallpaint = sf::Color::Red;    break;
    //         case 2:  wallpaint = sf::Color::Green;  break;
    //         case 3:  wallpaint = sf::Color::Blue;   break;
    //         case 4:  wallpaint = sf::Color::White;  break;
    //         default: wallpaint = sf::Color::Yellow; break;
    //     }

    //     if(side == 1) {
    //         wallpaint.r /= 2;
    //         wallpaint.g /= 2;
    //         wallpaint.b /= 2;
    //     }

    //     sw::Line wall;
    //         wall.setThickness(0.0f);
    //         wall.setPoint(wall.getStartIndex(), {x, drawStart});
    //         wall.setPoint(wall.getEndIndex()  , {x, drawEnd});
    //         wall.setColor(wallpaint);
    //     window->draw(wall);
    // }


    // std::cout << "PLX: " << playerX << " ; PLY: " << planeY << std::endl;
    // std::cout << "DRX: " << dirX << " ; DRY: " << dirY << std::endl;
    // std::cout << "PPX: " << planeX << " ; PPY: " << planeY << std::endl;
    // std::cout << "-------------------------------------------" << std::endl;

File: sfml-rust/src/player/ray.rs
use sfml::graphics::{Color, RenderTarget, Texture, IntRect, RcTexture, RenderStates, Text, Font, Sprite, RectangleShape, Shape};
use sfml::graphics::{RenderWindow, Transformable};
use sfml::system::Vector2f;

use std::f32::consts::PI;           // For PI   constant (f32)
use std::f32::consts::FRAC_PI_2;    // For PI/2 constant (f32)

use crate::radians;
use super::wline::WideLine;
use super::{Player, Wall};  // We want to use structs from from modules used in parent module

// use cpp:cpp;

pub struct Ray;     // Ray stores nothing
impl Ray {
    pub fn cast(f: f32, player: &Player, map: &Wall, window: &mut RenderWindow) {
        let mut dist = 0.0;      // Shortest distance to wall

        let width  = window.size().x;
        let height = window.size().y;

        let dof_range = 64;
        let mut dof: i32;               // Deepth of field
        let fov = f;                    // Field of view

        let mut ray_x = 0.0;            // Ray X position
        let mut ray_y = 0.0;            // Ray Y position
        let mut ray_ang: f32;           // Ray angle

        let cell = map.cell as i64;     // Cell size
        let map_w = map.width as i64;   // Map width
        let map_h = map.height as i64;  // Map height
        let mut map_tv = 0;             // Cell type on vertical axis (0, 1, 2, 3 etc.)
        let mut map_th = 0;             // Cell type on horizontal axis (0, 1, 2, 3 etc.)
        
        let mut map_x: i64;             // On what cell, on X axis, player is standing
        let mut map_y: i64;             // On what cell, on Y asix, player is standing
        let mut map_pos: i64;           // Index for map array

        let mut offset_x = 0.0;         // Ray offset on X axis
        let mut offset_y = 0.0;         // Ray offset on Y axis
        let grid = &map.grid;           // Our map

        let player_x = player.player.position().x;              // Player position on X axis in general
        let player_y = player.player.position().y;              // Player position on Y axis in general
        let player_ang = radians(player.player.rotation());     // Player angle (heading)



        // Difference between (map_x, map_y) and (player_x, player_y) is simple:
        // (0, 1) vs (436.11315, 313.21948)
        // (5, 6) vs (101.31454, 201.41210)
        // (map_x, map_y) are coordinates for map from `map.grid`
        // (player_x, player_y) are coordinates for position in the window

        // `ray_ang`  vs  `player_ang` is also simple
        // `player_ang` is heading - based on that we know if we're moving right, top etc.
        // `ray_ang` is an angle between -(fov/2) and (fov/2) - we're drawing rays between this range


        // Ray angle is in radians
        ray_ang = player_ang - radians(fov/2.0);
        if ray_ang < 0.0    {ray_ang = ray_ang + 2.0*PI;}
        if ray_ang > 2.0*PI {ray_ang = ray_ang - 2.0*PI;}

        for r in 0..width as i32 {
            // Horizontal line
            let mut dist_h = f32::INFINITY;     // We want some big number here for the start
            // let mut hor_x = player_x;
            // let mut hor_y = player_y;

            dof = 0;
            let a_tan = -1.0/f32::tan(ray_ang);


            if ray_ang > PI {
                ray_y = (((player_y as i64)/cell)*cell) as f32 - 0.0001;   // Rounding to the very edge of the wall
                offset_y = -cell as f32;
            }
            if ray_ang < PI {
                ray_y = ((((player_y as i64)/cell)*cell) + cell) as f32;
                offset_y = cell as f32;
            }
            if ray_ang == 0.0 || ray_ang == PI {
                // Lines here are parallel - they never meet 
                ray_x = player_x;
                ray_y = player_y;
                dof = dof_range;
            }

            ray_x = (player_y - ray_y) * a_tan + player_x;
            offset_x = -offset_y * a_tan;


            while dof < dof_range {
                map_x = (ray_x as i64)/cell;
                map_y = (ray_y as i64)/cell;
                map_pos = map_y * map_w + map_x;        // Index for map grid - that's why it is 1D array

                if (map_pos > 0) && (map_pos < map_w * map_h) && (grid[map_pos as usize] > 0) {
                    // We found the shortest path to horizontal wall
                    map_th = grid[map_pos as usize];    // Map has different kinds of cells
                    // hor_x = ray_x;
                    // hor_y = ray_y;
                    dist_h = Self::dist(player_x, player_y,  ray_x, ray_y);

                    dof = dof_range;
                } else {
                    // We're still searching
                    ray_x = ray_x + offset_x;
                    ray_y = ray_y + offset_y;
                    dof = dof + 1;
                }
            }


            // sprite.set_scale(Vector2f::new(texture_offset_x_hor as f32, texture_offset_y_hor as f32));
            // sprite.set_position(Vector2f::new(texture_offset_x_hor as f32, texture_offset_y_hor as f32));


            // Vertical line
            let mut dist_v = f32::INFINITY;     // We want some big number here for the start
            let mut vert_x = player_x;
            let mut vert_y = player_y;

            dof = 0;
            let n_tan = -f32::tan(ray_ang);

            
            if ray_ang > FRAC_PI_2 && ray_ang < 3.0*FRAC_PI_2 {
                ray_x = (((player_x as i64)/cell)*cell) as f32 - 0.0001;   // Rounding to the very edge of the wall
                offset_x = -cell as f32;
            }
            if ray_ang < FRAC_PI_2 || ray_ang > 3.0*FRAC_PI_2 {
                ray_x = ((((player_x as i64)/cell)*cell) + cell) as f32;
                offset_x = cell as f32;
                
            }
            if ray_ang == 0.0 || ray_ang == PI {
                // Lines here are parallel - they never meet 
                ray_x = player_x;
                ray_y = player_y;
                dof = dof_range;
            }

            ray_y = (player_x - ray_x) * n_tan + player_y;
            offset_y = -offset_x * n_tan;


            while dof < dof_range {
                map_x = (ray_x as i64)/cell;
                map_y = (ray_y as i64)/cell;
                map_pos = map_y * map_w + map_x;        // Index for map grid - that's why it is 1D array

                if (map_pos > 0) && (map_pos < map_w * map_h) && (grid[map_pos as usize] > 0) {
                    // We found the shortest path to horizontal wall
                    map_tv = grid[map_pos as usize];    // Map has different kinds of cells
                    // vert_x = ray_x;
                    // vert_y = ray_y;
                    dist_v = Self::dist(player_x, player_y,  ray_x, ray_y);

                    dof = dof_range;
                } else {
                    // We're still searching
                    ray_x = ray_x + offset_x;
                    ray_y = ray_y + offset_y;
                    dof = dof + 1;
                }
            }
            

            // sprite.set_scale(Vector2f::new(texture_offset_x_vert as f32, texture_offset_y_vert as f32));
            // sprite.set_position(Vector2f::new(texture_offset_x_vert as f32, texture_offset_y_vert as f32));


            // Distance based shading for 3D walls and colouring based on cell type
            let mut wallpaint = Color::rgb(0, 0, 0);
            let brightness: f32;
            let width_sqr = (width+width+width) as f32;    // Magic number - screen is 1024, but we're drawing only on the half of it 
            let dist_sqr: f32;

            // We only want to draw shortes ray from dist_v and dist_h
            if dist_v < dist_h {
                // ray_x = vert_x;
                // ray_y = vert_y;
                dist = dist_v;

                dist_sqr = dist*dist;
                brightness = Self::map(dist_sqr, 0.0, width_sqr, 255.0, 0.0);

                // Shading and slightly different from horizontal walls color
                match map_tv {
                    1 => {
                        wallpaint.r = (brightness/2.0) as u8;
                        wallpaint.g = (brightness/2.0) as u8;
                        wallpaint.b = (brightness/2.0) as u8;
                    },
                    2 => {wallpaint.r = brightness as u8;},
                    3 => {wallpaint.b = brightness as u8;},
                    4 => {wallpaint.r = brightness as u8; wallpaint.g = brightness as u8;},
                    _ => {wallpaint = Color::BLACK},
                }
            } else if dist_h < dist_v {
                // ray_x = hor_x;
                // ray_y = hor_y;
                dist = dist_h;

                dist_sqr = dist*dist;
                brightness = Self::map(dist_sqr, 0.0, width_sqr, 220.0, 0.0);
            
                match map_th {
                    1 => {
                        wallpaint.r = (brightness/2.0) as u8;
                        wallpaint.g = (brightness/2.0) as u8;
                        wallpaint.b = (brightness/2.0) as u8;
                    },
                    2 => {wallpaint.r = brightness as u8;},
                    3 => {wallpaint.b = brightness as u8;},
                    4 => {wallpaint.r = brightness as u8; wallpaint.g = brightness as u8;},
                    _ => {wallpaint = Color::BLACK;},
                }
            } else {brightness = 127.0; wallpaint.b = brightness as u8;}


            // let ray = WideLine::new(Vector2f::new(player_x, player_y), Vector2f::new(ray_x, ray_y), 1.0, wallpaint);
            // ray.draw(window);
            
            ray_ang = ray_ang + radians(fov / width as f32);   // Magic number - offset each ray by 1 degree

            if ray_ang < 0.0    {ray_ang = ray_ang + 2.0*PI;}
            if ray_ang > 2.0*PI {ray_ang = ray_ang - 2.0*PI;}


            // 3D walls
            // Removes fish eye effect
            let mut cell_ang = player_ang - ray_ang;
            if cell_ang < 0.0    {cell_ang = cell_ang + 2.0*PI;}
            if cell_ang > 2.0*PI {cell_ang = cell_ang - 2.0*PI;}
            dist = dist * f32::cos(cell_ang);


            let line_height = (cell*height as i64) as f32 / dist;   // Walls height - can be regulated
            let wall_width = 1;                                     // Space taken on the screen by single strip
            let line_offset = (height/2) as f32 - line_height/2.0;  // Camera height
            let wall_ofsset = 1.0;                                  // If we want to draw map and walls on the same time

            let wall = WideLine::new(Vector2f::new(((r as i32) * wall_width) as f32 + wall_ofsset, line_offset),                // FROM where
                                     Vector2f::new(((r as i32) * wall_width) as f32 + wall_ofsset, line_height+line_offset),    // TO   where
                                     wall_width as f32, wallpaint);
            wall.draw(window);


            // sprite.set_texture_rect(IntRect::new((r * wall_width) + (wall_ofsset as i32), line_offset as i32,  1,  (line_height + line_offset) as i32));
            // sprite.set_texture_rect(IntRect::new(0, 0,  1, 64));
            // sprite.set_texture_rect(IntRect::new(0, 0,  1, (line_height) as i32));
            // sprite.set_scale(Vector2f::new(1.0, line_height));
            
            // sprite.set_position(Vector2f::new(513.0, 0.0));
            
            
            // sprite.set_position(Vector2f::new(((r * wall_width) + (wall_ofsset as i32)) as f32, line_offset));
            // sprite.set_scale(Vector2f::new(1.0/cell as f32, line_height/cell as f32));
            // sprite.set_scale(Vector2f::new(1.0/cell as f32, line_height/cell as f32));

            // window.draw(&sprite);
        }
    }

    // Distance between two points
    fn dist(ax: f32, ay: f32, bx: f32, by: f32) -> f32 {
        f32::sqrt(f32::powi(ax-bx, 2) + f32::powi(ay-by, 2))
    }

    // Map one range to another
    fn map(num: f32, in_min: f32, in_max: f32, out_min: f32, out_max: f32) -> f32 {
        (num-in_min) * (out_max-out_min)/(in_max-in_min) + out_min
    }
}